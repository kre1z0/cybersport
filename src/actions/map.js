import {SET_CENTER, RESET_MAP, SET_RESOLUTION} from '../reducers/map';

export const setCenter = (center) => ({
    type: SET_CENTER,
    center
});

export const resetMap = () => ({
    type: RESET_MAP
});

export const setResolution = (resolution) => ({
    type: SET_RESOLUTION,
    resolution
});
