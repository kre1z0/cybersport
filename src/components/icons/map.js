import React from 'react';
import SvgIcon from 'material-ui/SvgIcon';
import {silver, strongSilver, softGreen} from '../../assets/theme';

const MapIcon = ({isActive, ...props}) => (
    <SvgIcon {...props} viewBox="0 0 18 23">
        <g>
            <path fill={isActive ? softGreen : silver} d="M9,10.3c-1.7,0-3-1.4-3-3s1.4-3,3-3s3,1.4,3,3S10.7,10.3,9,10.3z M9,6.3c-0.6,0-1,0.5-1,1s0.5,1,1,1
			s1-0.5,1-1S9.6,6.3,9,6.3z"/>
            <path fill={isActive ? softGreen : silver} d="M15,11.7c1-1.3,1.5-2.8,1.5-4.4c0-4-3.4-7.3-7.5-7.3S1.5,3.3,1.5,7.3c0,1.6,0.5,3.1,1.5,4.4c0,0,0,0,0,0l0,0
			c0,0,0,0,0,0c0,0,0,0,0,0l5.2,6.9c0,0,0.1,0,0.1,0.1c0,0,0.1,0.1,0.1,0.1c0,0,0.1,0,0.1,0.1c0,0,0.1,0,0.1,0.1C8.7,19,8.9,19,9,19
			c0,0,0,0,0,0s0,0,0,0c0.1,0,0.3,0,0.4-0.1c0,0,0.1,0,0.1-0.1c0,0,0.1,0,0.1-0.1c0,0,0.1-0.1,0.1-0.1c0,0,0.1,0,0.1-0.1L15,11.7
			C15,11.7,15,11.7,15,11.7C15,11.7,15,11.7,15,11.7L15,11.7C15,11.7,15,11.7,15,11.7z M9,2c3,0,5.5,2.4,5.5,5.3
			c0,1.2-0.4,2.2-1.1,3.2c0,0,0,0,0,0c0,0,0,0,0,0L9,16.3l-4.4-5.8c0,0,0,0,0,0c0,0,0,0,0,0c-0.7-0.9-1.1-2-1.1-3.2C3.5,4.4,6,2,9,2
			z"/>
        </g>
        <path fill={isActive ? strongSilver : silver} d="M17,23H1c-0.6,0-1-0.4-1-1s0.4-1,1-1h16c0.6,0,1,0.4,1,1S17.6,23,17,23z"/>
    </SvgIcon>
);

export default MapIcon;


