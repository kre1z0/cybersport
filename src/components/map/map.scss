.map-wrapper{
  height: 100%;
  width: 100%;
  button{
    color: #0074D9;
  }
}