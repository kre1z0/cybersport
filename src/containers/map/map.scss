:global {
  .map-container {
    height: 100%;
    width: 100%;
    position: relative;
  }
  .sGis-zoomButtons{
    position: absolute;
    top: calc( 50% - 54px );
    left: 10px;
    .sGis-zoomButtons-zoomIn, .sGis-zoomButtons-zoomOut{
      cursor: pointer;
      margin: 2px;
      z-index: 9;
      position: relative;
      width: 44px;
      height: 44px;
      background-color: #ffffff;
      border-radius: 50%;
      background-repeat: no-repeat;
      background-size: 16px;
      background-position: center;
    }
    .sGis-zoomButtons{
      &-zoomIn{
        background-image: url(../../assets/images/plus.svg);
        margin-bottom: 10px;
      }
      &-zoomOut{
        background-image: url(../../assets/images/minus.svg);
      }
    }
  }
}