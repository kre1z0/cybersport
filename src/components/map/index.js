import React, {Component, PropTypes} from 'react';
import sGis from '../../assets/sgis';
import debounce from 'lodash/debounce';

import './map.css';

class Map extends Component {
    static propTypes = {
        center: PropTypes.array,
        resolution: PropTypes.number,
        onCenterChange: PropTypes.func,
        onResolutionChange: PropTypes.func
    };
    static defaultProps = {
        center: [0, 0],
        resolution: 76.437
    };
    
    onResolutionChange = (resolution) => {
        if (resolution !== this.props.resolution) {
            this.props.onResolutionChange(resolution);
        }
    };
    
    onCenterChange = (position) => {
        const newCenter = new sGis.Point(position, sGis.CRS.webMercator);
        const oldCenter = new sGis.Point(this.props.center, sGis.CRS.webMercator);
        if (!newCenter.equals(oldCenter)) {
            this.props.onCenterChange(newCenter.position);
        }
    };
    
    componentDidMount () {
        this.sgis = sGis.init({
            wrapper: this.container,
            position: this.props.center,
            resolution: 76.437,
            layers: [ new sGis.TileLayer('http://b.tile.openstreetmap.org/{z}/{x}/{y}.png') ]
        });
        
        this.sgis.map.on('bboxChange', debounce(({sourceObject: {resolution, position}})=>{
                this.onResolutionChange(resolution);
                this.onCenterChange(position);
        }, 500))
    }
    
    componentWillReceiveProps (nextProps) {
        const newCenter = new sGis.Point(nextProps.center, sGis.CRS.webMercator);
        
        if (!newCenter.equals(this.sgis.map.centerPoint)) {
            this.sgis.map.position = newCenter.position;
        }
        
        if (nextProps.resolution !== this.sgis.map.resolution) {
            this.sgis.map.resolution = nextProps.resolution;
        }
    }
    
    componentWillUnmount () {
        delete this.sgis;
    }
    
    render () {
        return (
            <div className="map-container" ref={container => this.container = container}>
    
            </div>
        );
    }
}

export default Map