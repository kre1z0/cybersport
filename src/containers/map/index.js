/**
 * Created by tporyadin on 3/6/2017.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Loader from 'material-ui/CircularProgress';

import {
    setCenter,
    setResolution,
    setObjectsDataFilter,
    loadMapServices,
    pickObject,
} from '../../ducks/map';
import getLayerManager, { isServicesLoaded } from '../../evergis/layer-manager';
import { setDataFilters } from '../../evergis/api';

import Map from '../../components/map';
import LayersList from './layer-list';
import FeaturePopup from './feature-popup';

import './map.scss';

class MapContainer extends Component {
    static propTypes = {
        map: PropTypes.object,
        setCenter: PropTypes.func,
        setResolution: PropTypes.func,
        setObjectsDataFilter: PropTypes.func,
        loadMapServices: PropTypes.func,
        isAuth: PropTypes.bool,
    };

    componentDidMount() {
        const { loadMapServices, isAuth } = this.props;
        if (isAuth && !isServicesLoaded()) {
            loadMapServices().then(services => {
                this.updateServices(services, this.props.map);
            });
        }
    }

    componentWillReceiveProps({ loadMapServices, map, isAuth }) {
        if (
            !this.props.isAuth &&
            isAuth &&
            !map.loading &&
            !isServicesLoaded()
        ) {
            loadMapServices().then(services => {
                this.updateServices(services, map);
            });
        }

        let checkStatus = false;
        for (let i in map.objectsDataFilter) {
            if (
                this.props.map.objectsDataFilter[i] !== map.objectsDataFilter[i]
            )
                checkStatus = true;
        }
        if (checkStatus && isAuth) {
            const layerManager = getLayerManager();
            const sber_service = layerManager.getService('sber_objects');
            let df = sber_service.dataFilter;
            df.condition = this.setFilter(map.objectsDataFilter);
            sber_service.setDataFilter(df);
        }
    }

    updateServices(services, map) {
        const layerManager = getLayerManager();
        services.forEach(({ name, isVisible }) => {
            const service = layerManager.getService(name);

            if (name === 'sber_objects') {
                let df = service.dataFilter;
                df.condition = this.setFilter(map.objectsDataFilter);
                service.setDataFilter(df);
            }

            if (service) {
                service.isDisplayed = isVisible;
            }
        });
    }

    setFilter(filter) {
        let filterArray = [];
        let toServiceFilter = null;
        for (let f in filter) {
            if (!filter[f]) filterArray.push('status != ' + f);
        }
        if (filterArray.length > 0) {
            toServiceFilter = filterArray.join(' && ');
        }
        return toServiceFilter;
    }

    onMapPick = e => {
        this.props.pickObject(e.point);
    };

    render() {
        const {
            map,
            setCenter,
            setResolution,
            setObjectsDataFilter,
            isAuth,
        } = this.props;

        return (
            <div className="map-container">
                {isAuth
                    ? <Map
                          center={map.center}
                          resolution={map.resolution}
                          onCenterChange={setCenter}
                          onResolutionChange={setResolution}
                          onMapPick={this.onMapPick}
                      />
                    : <Loader className="loader" />}
                <LayersList
                    objectsDataFilter={map.objectsDataFilter}
                    onChangeItem={setObjectsDataFilter}
                />
                <FeaturePopup />
            </div>
        );
    }
}

const mapProps = ({ map, user: { employee_id } }) => ({
    map,
    isAuth: !!employee_id,
});

const mapActions = {
    setCenter,
    setResolution,
    setObjectsDataFilter,
    loadMapServices,
    pickObject,
};

export default connect(mapProps, mapActions)(MapContainer);
