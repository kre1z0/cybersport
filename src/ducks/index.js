import { combineReducers } from 'redux-immutable';
import { Record } from 'immutable';

import map from './map';
import user from './user';
import plan from './plan';
import objects from './objects';
import domains from './domains';
import employees from './employees';
import inspections from './inspections';

export const StateRecord = new Record({
    user: undefined,
    plan: undefined,
    map: undefined,
    objects: undefined,
    domains: undefined,
    employees: undefined,
    inspections: undefined,
});

export default combineReducers({
    map,
    user,
    plan,
    objects,
    domains,
    employees,
    inspections,
});
