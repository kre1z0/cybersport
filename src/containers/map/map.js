/**
 * Created by tporyadin on 3/6/2017.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import withAuth from '../../hoc/withAuth';

import {
    setCenter,
    setResolution,
    loadMapServicesIfNeeded,
    selectObject,
} from '../../ducks/map';
import { getDomainsIfNeeded } from '../../ducks/domains';
import getLayerManager from '../../evergis/layer-manager';
import { getFeatureLayer } from '../../evergis/map';

import Map from '../../components/map';
import LayersList from './layer-list';
import FeaturePopup from './feature-popup';

import FloatingActionButton from 'material-ui/FloatingActionButton';

import { MapLayers } from '../../components/icons';

import {
    OBJECTS_SERVICE,
    OFFICES_SERVICE,
    EMPLOYEES_SERVICE,
} from '../../evergis/helpers';

import './map.scss';

const floatButtonStyles = {
    button: {
        position: 'absolute',
        top: '1rem',
        left: '1rem',
        zIndex: 1,
    },
    svg: { height: '15px', width: '16px', verticalAlign: 'sub' },
};

class MapContainer extends Component {
    state = {
        showPopup: false,
    };

    static propTypes = {
        map: PropTypes.object,
        setCenter: PropTypes.func,
        setResolution: PropTypes.func,
        setObjectsDataFilter: PropTypes.func,
        loadMapServices: PropTypes.func,
        isAuth: PropTypes.bool,
    };

    componentDidMount() {
        const { loadMapServicesIfNeeded, getDomainsIfNeeded, map } = this.props;
        loadMapServicesIfNeeded().then(services => {
            services && this.updateServices(services, map);
        });
        getDomainsIfNeeded();
    }

    componentWillReceiveProps({ map }) {
        if (
            this.props.map.objectsDataFilter !== map.objectsDataFilter ||
            this.props.map.domainsFilter !== map.domainsFilter
        ) {
            const layerManager = getLayerManager();
            const sber_service = layerManager.getService(OBJECTS_SERVICE);
            let df = sber_service.dataFilter;
            df.condition = this.getFilter(
                map.objectsDataFilter.toJS(), //todo
                map.domainsFilter,
            );
            sber_service.setDataFilter(df);
            sber_service.layer.redraw();
        }

        if (this.props.map.showOffices !== map.showOffices) {
            const offices_layer = getFeatureLayer(OFFICES_SERVICE);
            offices_layer.isDisplayed = map.showOffices; //todo
        }

        if (this.props.map.showHomeAddress !== map.showHomeAddress) {
            const employess_layer = getFeatureLayer(EMPLOYEES_SERVICE);
            employess_layer.isDisplayed = map.showHomeAddress; //todo
        }
    }

    updateServices(services, map) {
        const layerManager = getLayerManager();
        services.forEach(({ name, isVisible }) => {
            const service = layerManager.getService(name);

            if (name === OBJECTS_SERVICE) {
                let df = service.dataFilter;
                df.condition = this.getFilter(
                    map.objectsDataFilter.toJS(),
                    map.domainsFilter,
                );
                service.setDataFilter(df);
                service.layer.redraw();
            }

            if (service) {
                service.isDisplayed = isVisible;
            }
        });
    }

    getFilter(filter, domains) {
        let filterArray = [];
        let toServiceFilter = null;
        for (let f in filter) {
            if (!filter[f]) filterArray.push('status != ' + f);
        }
        if (filterArray.length > 0) {
            toServiceFilter = filterArray.join(' && ');
        }
        return (
            (toServiceFilter &&
                domains &&
                toServiceFilter + ' && ' + domains) ||
            toServiceFilter ||
            domains
        );
    }

    onMapClick = e => {
        this.props.selectObject({ objects: [] });
    };

    handleShowPopup = () => {
        this.setState(state => ({
            showPopup: !state.showPopup,
        }));
    };

    render() {
        const { map, setCenter, setResolution } = this.props;
        const { showPopup } = this.state;

        return (
            <div className="map-container">
                <Map
                    center={map.center}
                    resolution={map.resolution}
                    onCenterChange={setCenter}
                    onResolutionChange={setResolution}
                    onMapClick={this.onMapClick}
                />
                <FloatingActionButton
                    style={floatButtonStyles.button}
                    backgroundColor="#fff"
                    onTouchTap={this.handleShowPopup}
                >
                    <MapLayers
                        style={floatButtonStyles.svg}
                        isActive={showPopup}
                    />
                </FloatingActionButton>
                {showPopup && <LayersList />}
                <FeaturePopup />
            </div>
        );
    }
}

const mapProps = ({ map }) => ({
    map,
});

const mapActions = {
    setCenter,
    setResolution,
    loadMapServicesIfNeeded,
    selectObject,
    getDomainsIfNeeded,
};

export default connect(mapProps, mapActions)(withAuth(MapContainer));
