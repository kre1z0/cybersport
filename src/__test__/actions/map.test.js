import { setCenter, resetMap, setResolution } from '../../ducks/map';

describe('map actions', () => {
    it('should create an action to set center', () => {
        expect(setCenter([0, 0])).toMatchSnapshot();
        expect(setCenter([500, 100])).toMatchSnapshot();
        expect(setCenter([100, 500])).toMatchSnapshot();
    });
    
    it('should create an action to set resolution', () => {
        expect(setResolution(0)).toMatchSnapshot();
        expect(setResolution(5)).toMatchSnapshot();
        expect(setResolution(100)).toMatchSnapshot();
    });
    
    it('should create an action to reset center', () => {
        expect(resetMap()).toMatchSnapshot();
    })
});
