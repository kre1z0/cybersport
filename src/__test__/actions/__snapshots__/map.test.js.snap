// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map actions should create an action to reset center 1`] = `
Object {
  "payload": undefined,
  "type": "[3] map/reset-map",
}
`;

exports[`map actions should create an action to set center 1`] = `
Object {
  "payload": Array [
    0,
    0,
  ],
  "type": "[2] map/set-center",
}
`;

exports[`map actions should create an action to set center 2`] = `
Object {
  "payload": Array [
    500,
    100,
  ],
  "type": "[2] map/set-center",
}
`;

exports[`map actions should create an action to set center 3`] = `
Object {
  "payload": Array [
    100,
    500,
  ],
  "type": "[2] map/set-center",
}
`;

exports[`map actions should create an action to set resolution 1`] = `
Object {
  "payload": 0,
  "type": "[4] map/set-resolution",
}
`;

exports[`map actions should create an action to set resolution 2`] = `
Object {
  "payload": 5,
  "type": "[4] map/set-resolution",
}
`;

exports[`map actions should create an action to set resolution 3`] = `
Object {
  "payload": 100,
  "type": "[4] map/set-resolution",
}
`;
